core/src/main/java/org/infinispan/iteration/impl/LocalEntryRetriever.java

```diff
@@ -246,6 +247,104 @@ protected boolean shouldUseLoader(Set<Flag> flags) {
+                        clone = entryFactory.create(marshalledKey, marshalledValue, entry);
+                     } else {
+                        clone = entry;
+                     }
+                     if (filter != null) {
+                        K key = clone.getKey();
+                        if (usedConverter == null && filter instanceof KeyValueFilterConverter) {
+                           C converted = ((KeyValueFilterConverter<K, V, C>)filter).filterAndConvert(
+                                 key, clone.getValue(), clone.getMetadata());
+                           if (converted != null) {
+                              // If entry and clone are the same we didn't have a marshalled value then we need to create a 
+                              // copy with the new value
+                              if (entry == clone) {
+                                 clone = (InternalCacheEntry<K, V>) entryFactory.create(clone.getKey(), converted, entry);
+                              } else {
+                                 clone.setValue((V) converted);
+                              }
+                           } else {
+                              continue;
+                           }
+                        }
+                        else if (!filter.accept(key, clone.getValue(), clone.getMetadata())) {
+                           continue;
+                        }
+                     }
+
+                     next = (CacheEntry<K, C>) clone;
+                     if (usedConverter != null) {
+                        C newValue = usedConverter.convert(clone.getKey(), clone.getValue(), clone.getMetadata());
+                        if (newValue != clone.getValue()) {
+                           // If we didn't have a marshalled value then we need to create a copy with the new value
+                           if (entry == clone) {
+                              next = entryFactory.create(clone.getKey(), newValue, entry);
+                           } else {
+                              next.setValue(newValue);
+                           }
+                        }
+                     }
+                  } else {
+                     break;
+                  }
+               }
+               return next != null;
+            }
+
+            @Override
+            public CacheEntry<K, C> next() {
+               if (next == null && !hasNext()) {
+                  throw new NoSuchElementException();
+               }
+               CacheEntry<K, C> returnEntry = next;
+               next = null;
+               prev = returnEntry;
+               return returnEntry;
+            }
+
+            @Override
+            public void remove() {
+               cache.remove(prev.getKey());
+            }
+
+            @Override
+            public void close() {
+               // Nothing to do here
+            }
+            
+         };
+      }
```

Comment:
retrieveEntries() is already huge, can't this code be refactored in submethods please ?